# ACTIVIDAD 7

**Maestría en Inteligencia Artificial y Analítica de Datos**

**Procesamiento de Lenguaje Natural**

Actividad 7: Ejercicios de representación de texto.

Objetivo de la actividad.
Prácticar las principales técnicas de representación de texto y visualizer las diferencias que se obtienen con sus variantes.

Instrucciones

Considera el pequeño corpus de reseñas de películas y comentarios de redes sociales que
está al final del documento y realiza los siguientes ejercicios:

Ejercicio 1: Utiliza el texto crudo (sin procesar)

* Crear un CountVectorizer y un TfidfVectorizer.
* Aplícalos al corpus y transformar los textos.
* Mostrar:
* Dimensiones de la matriz resultante.
* Primeros 10 tokens y sus índices.
* Comparar las diferencias entre ambas matrices (CountVectorizer y TfidfVectorizer),mencionando de manera general las diferencias observadas.

Ejercicio 2: Preprocesa el texto eliminando los acentos y stopwords, así como
convirtiéndolo a minúsculas.

* Realiza las mismas actividades que en el ejercicio anterior.
* Compara las matrices del ejercicio anterior con las matrices que obtuviste en este ejercicio, mencionando de manera general las diferencias observadas.

corpus_es = [
 "Me encantó la película, los actores fueron increíbles",
 "No me gustó la película, el guion fue muy malo",
 "La actuación fue excelente, pero la historia era predecible",
 "Película aburrida, me dormí a la mitad",
 "¡Maravillosa! Recomiendo esta película a todos mis amigos",
 "El guion estaba mal escrito, pero la actuación salvó la película",
 "Demasiado larga y lenta, no la volvería a ver",
 "Me gustó mucho, los efectos especiales fueron impresionantes",
 "No es mala, pero esperaba algo más emocionante",
 "Una obra maestra del cine español, increíble dirección",
 "El final fue confuso, no entendí nada",
 "Excelente cinematografía, pero el guion flojo",
 "Demasiado predecible, ya sabía lo que iba a pasar",
 "Me reí mucho, muy divertida y entretenida",
 "La música fue espectacular, pero los actores no convencieron",
 "Película mediocre, no aporta nada nuevo",
 "Excelente historia, emocionante hasta el final",
 "No me gustó la ambientación, parecía de bajo presupuesto",
 "Muy buena dirección y fotografía, la recomiendo",
 "El guion tenía agujeros, pero los efectos visuales impresionan",
 "Aburrida, diálogo poco natural",
 "Gran actuación de los protagonistas, realmente me impactó",
 "El ritmo es lento, pero la historia es interesante",
 "No la recomiendo, desperdicié mi tiempo",
 "Un clásico moderno, me encantó cada escena",
 "Las escenas de acción fueron espectaculares",
 "El humor es pobre y los personajes poco creíbles",
 "Me emocioné, la historia me llegó al corazón",
 "La trama es confusa y difícil de seguir",
 "Película fantástica, muy bien lograda"
]

# Corpus

In [ ]:
corpus_es = [
 "Me encantó la película, los actores fueron increíbles",
 "No me gustó la película, el guion fue muy malo",
 "La actuación fue excelente, pero la historia era predecible",
 "Película aburrida, me dormí a la mitad",
 "¡Maravillosa! Recomiendo esta película a todos mis amigos",
 "El guion estaba mal escrito, pero la actuación salvó la película",
 "Demasiado larga y lenta, no la volvería a ver",
 "Me gustó mucho, los efectos especiales fueron impresionantes",
 "No es mala, pero esperaba algo más emocionante",
 "Una obra maestra del cine español, increíble dirección",
 "El final fue confuso, no entendí nada",
 "Excelente cinematografía, pero el guion flojo",
 "Demasiado predecible, ya sabía lo que iba a pasar",
 "Me reí mucho, muy divertida y entretenida",
 "La música fue espectacular, pero los actores no convencieron",
 "Película mediocre, no aporta nada nuevo",
 "Excelente historia, emocionante hasta el final",
 "No me gustó la ambientación, parecía de bajo presupuesto",
 "Muy buena dirección y fotografía, la recomiendo",
 "El guion tenía agujeros, pero los efectos visuales impresionan",
 "Aburrida, diálogo poco natural",
 "Gran actuación de los protagonistas, realmente me impactó",
 "El ritmo es lento, pero la historia es interesante",
 "No la recomiendo, desperdicié mi tiempo",
 "Un clásico moderno, me encantó cada escena",
 "Las escenas de acción fueron espectaculares",
 "El humor es pobre y los personajes poco creíbles",
 "Me emocioné, la historia me llegó al corazón",
 "La trama es confusa y difícil de seguir",
 "Película fantástica, muy bien lograda"
]

#INSTRUCCIONES:
##Ejercicio 1: Utiliza el texto crudo (sin procesar)
* Crear un CountVectorizer y un TfidfVectorizer.
* Aplícalos al corpus y transformar los textos.
### Mostrar:
* Dimensiones de la matriz resultante.
* Primeros 10 tokens y sus índices.
* Comparar las diferencias entre ambas matrices (CountVectorizer y TfidfVectorizer),mencionando de manera general las diferencias observadas.


In [ ]:
def separador():
    print("\n" + "=" * 170 + "\n")
def header(Titulo):
  por=((170-len(Titulo))//10)
  a=int(1.2*por)
  print("."*a+"·"*a+"~"*a+"≈"*a+"≋"
  *int(por*.2),Titulo,"≋"*int(por*.2)
  +"≈"*a+"~"*a+"·"*a+"."*a,"\n\n")

In [ ]:
#se importan librerias
import pandas as pd #pandas para crear el df
from sklearn.feature_extraction.text import CountVectorizer # importacion de countvectorizer (BoW)
#se importa tdidfVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer
import nltk #se importa nltk para stopwords
from nltk.corpus import stopwords #se importa stopwords
nltk.download('stopwords') #se importan stopwords en español

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [ ]:
stopwords=stopwords.words('spanish')
#función para la transformación del texto en vectores para countvectorizer:
def countvectorizer(corpus, lower_case, stopwords, stripacc):
  vectorizer=CountVectorizer(lowercase=lower_case, stop_words=stopwords, strip_accents=stripacc)#todo en crudo, sin hiperparámetros
  x=vectorizer.fit_transform(corpus)

  #se convierte a df para visualizar, y los nombres con get_feature_names_out

  bag_df_p=pd.DataFrame(data=x.toarray(), columns= vectorizer.get_feature_names_out())
  bag_df=bag_df_p.T #Se transpone la matriz para mostrar los tokens como filas
  for nombre, contenido in globals().items():
    if contenido is corpus:
      print(f'Tamaño de la matriz resultante del texto "{nombre}" con CountVectorizer:\nFilas:{bag_df.shape[0]}\nColumnas:{bag_df.shape[1]}')
      print(f'Primeros 10 tokens:\n{bag_df.head(10).to_string()}');separador()
#=============================================================================================================================================#
#Función para la transformación del texto en vectores con TF-IDF.
def tfidfvectorizer(corpus, lower_case, stopwords, stripacc):
  vectorizer=TfidfVectorizer(lowercase=lower_case, stop_words=stopwords, strip_accents=stripacc)#sin hiperparámetros
  tf=vectorizer.fit_transform(corpus)

  #se convierte a df para visualizar, y los nombres con get_feature_names_out
  tf_df=pd.DataFrame(tf.toarray(), columns=vectorizer.get_feature_names_out())
  tf_df=tf_df.T #Se transpone la matriz para mostrar los tokens como filas
  for nombre, contenido in globals().items():
    if contenido is corpus:
      print(f'Tamaño de la matriz resultante del texto "{nombre}" con TF-IDF:\nFilas:{tf_df.shape[0]}\nColumnas:{tf_df.shape[1]}')
      print(f'Primeros 10 tokens:\n{tf_df.head(10).to_string()}\n');separador()

In [ ]:
countvectorizer(corpus_es, False, None, None)
tfidfvectorizer(corpus_es, False, None, None)

Tamaño de la matriz resultante del texto "corpus_es" con CountVectorizer:
Filas:136
Columnas:30
Primeros 10 tokens:
             0   1   2   3   4   5   6   7   8   9   10  11  12  13  14  15  16  17  18  19  20  21  22  23  24  25  26  27  28  29
Aburrida      0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   1   0   0   0   0   0   0   0   0   0
Demasiado     0   0   0   0   0   0   1   0   0   0   0   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
El            0   0   0   0   0   1   0   0   0   0   1   0   0   0   0   0   0   0   0   1   0   0   1   0   0   0   1   0   0   0
Excelente     0   0   0   0   0   0   0   0   0   0   0   1   0   0   0   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0
Gran          0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   1   0   0   0   0   0   0   0   0
La            0   0   1   0   0   0   0   0   0   0   0   0   0   0   1   0   0   0   0   0 

**Comparar las diferencias entre ambas matrices (CountVectorizer y TfidfVectorizer),mencionando de manera general las diferencias observadas.**

Comienzo por una similitud, ambas matrices son del mismo tamaño (porque sin hiperparametros presentan la misma cantidad de tokens). En cuanto a las diferencias, CountVectorizer muestra el número de ocurrencias de la palabra dentro de una frase (documento) y tf-idf muestra la relevancia de la palabra dentro del texto usando un valor flotante, utilizando para ello el cuerpo completo de frases. En este algoritmoas palabras muy frecuentes en todos los documentos no son importantes porque no ayudan a diferenciar, en cambio bag of words no hace esa distinción.

En conclusión, CountVectorizer sirve solo para analizar la frecuencia, en cambio df-idf destaca lo distintivo, lo que lo hace bueno para clasificacion por ejemplo.

#======================================================================================




##Ejercicio 2:
 Preprocesa el texto eliminando los acentos y stopwords, así como
convirtiéndolo a minúsculas.
* Realiza las mismas actividades que en el ejercicio anterior.


In [ ]:
#vectorización del texto con las librerias utilizadas anteriormente, preprocesando el texto quitando los acentos, stopwords y en minusculas.
countvectorizer(corpus_es, True, stopwords, 'unicode')
tfidfvectorizer(corpus_es, True, stopwords, 'unicode')


Tamaño de la matriz resultante del texto "corpus_es" con CountVectorizer:
Filas:97
Columnas:30
Primeros 10 tokens:
              0   1   2   3   4   5   6   7   8   9   10  11  12  13  14  15  16  17  18  19  20  21  22  23  24  25  26  27  28  29
aburrida       0   0   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   1   0   0   0   0   0   0   0   0   0
accion         0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   1   0   0   0   0
actores        1   0   0   0   0   0   0   0   0   0   0   0   0   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
actuacion      0   0   1   0   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   1   0   0   0   0   0   0   0   0
agujeros       0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   1   0   0   0   0   0   0   0   0   0   0
ambientacion   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   1   

/usr/local/lib/python3.12/dist-packages/sklearn/feature_extraction/text.py:402: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ['eramos', 'estabamos', 'estais', 'estan', 'estara', 'estaran', 'estaras', 'estare', 'estareis', 'estaria', 'estariais', 'estariamos', 'estarian', 'estarias', 'esteis', 'esten', 'estes', 'estuvieramos', 'estuviesemos', 'fueramos', 'fuesemos', 'habeis', 'habia', 'habiais', 'habiamos', 'habian', 'habias', 'habra', 'habran', 'habras', 'habre', 'habreis', 'habria', 'habriais', 'habriamos', 'habrian', 'habrias', 'hayais', 'hubieramos', 'hubiesemos', 'mas', 'mia', 'mias', 'mio', 'mios', 'seais', 'sera', 'seran', 'seras', 'sere', 'sereis', 'seria', 'seriais', 'seriamos', 'serian', 'serias', 'si', 'tambien', 'tendra', 'tendran', 'tendras', 'tendre', 'tendreis', 'tendria', 'tendriais', 'tendriamos', 'tendrian', 'tendrias', 'teneis', 'tengais', 'tenia', 'teniais', 'teniamos', 'tenian', 'tenias', 'tuvie

* Compara las matrices del ejercicio anterior con las matrices que obtuviste en este ejercicio, mencionando de manera general las diferencias observadas.




Las matrices sin preprocesado fueron de:

* Filas:136
* Columnas:30

Y las posteriores, luego del preprocesado fueron de:

* Filas:97
* Columnas:30

Como era de esperarse, al quitar los stopwords se disminuye drasticamente el tamaño de la matriz, porque los tokens son menos. Las columnas (vocabulario resultante) permanecen del mismo tamaño, ya que el texto de origen es el mismo. También se unificaron palabras que en un principio aparecían con mayuscula y minúscula, por ejemplo "Perro" y "perro", esos dos terminos se convierten en uno solo al minimizar todos los tokens. La remoción de acentos podría influir de la misma forma al unificar palabras con alguna falta ortográfica (falta de acentuación por ej.) en algunos términos.
